java/fury-core/src/main/java/org/apache/fury/resolver/ClassInfo.java

```diff
@@ -86,14 +85,19 @@ public class ClassInfo {
     } else {
       this.fullClassNameBytes = null;
     }
+    // When `classId == ClassResolver.REPLACE_STUB_ID` was established,
+    // means only classes are serialized, not the instance. If we
+    // serialize such class only, we need to write classname bytes.
     if (cls != null
-        && (classId == ClassResolver.NO_CLASS_ID || classId == ClassResolver.REPLACE_STUB_ID)) {
+        && ((classId == ClassResolver.NO_CLASS_ID
+                && !classResolver.getFury().getConfig().isMetaShareEnabled())
+            || classId == ClassResolver.REPLACE_STUB_ID)) {
       // REPLACE_STUB_ID for write replace class in `ClassSerializer`.
       Tuple2<String, String> tuple2 = Encoders.encodePkgAndClass(cls);
       this.packageNameBytes =
-          metaStringResolver.getOrCreateMetaStringBytes(PACKAGE_ENCODER.encode(tuple2.f0));
+          metaStringResolver.getOrCreateMetaStringBytes(Encoders.encodePackage(tuple2.f0));
       this.classNameBytes =
-          metaStringResolver.getOrCreateMetaStringBytes(TYPE_NAME_ENCODER.encode(tuple2.f1));
+          metaStringResolver.getOrCreateMetaStringBytes(Encoders.encodeTypeName(tuple2.f1));
```

Comment:
Could you check whether we need to update MetaStringBytes.decode

Reply:
Updated.

`MetaStringBytes.decode` can also reuse existing `MetaStringDecoder` objects

Reply:
We can also create a `Decoders` class, similar to `org.apache.fury.meta.Encoders`, which can be used to cache `ConcurrentMap<MetaString, String>`, merge this PR first, and then I will do this thing.

Reply:
I prefer put it into Encoders instead

Reply:
> Updated.
> 
> `MetaStringBytes.decode` can also reuse existing `MetaStringDecoder` objects

You should update MetaStringBytes#decode too:
```java
  public String decode(MetaStringDecoder decoder) {
    int header = (int) (hashCode & 0xff);
    int encodingFlags = header & 0b111;
    MetaString.Encoding encoding = MetaString.Encoding.values()[encodingFlags];
    return decoder.decode(bytes, encoding);
  }
```
Currently it use the whole encoding set, but `Encoders.encodePackage` only use part of the encodings. The decode error will happen if a classname with unicode chars occur.

Could you add a unit test to cover such cases? It should raise error in your current change.

And the key here is that MetaStringBytes can't distinguish whether the data is  package or classname, so the decode need use the whole encoding set

Reply:
I think we should not update `MetaStringBytes#decode` for the following reasons:

`Encoders.encodePackage` supports UTF-8 encoding by default, so even if the classname contains unicode characters, it can be encoded successfully.

```java
static final Encoding[] pkgEncodings =
      new Encoding[] {UTF_8, ALL_TO_LOWER_SPECIAL, LOWER_UPPER_DIGIT_SPECIAL};

static final Encoding[] typeNameEncodings =
    new Encoding[] {
      UTF_8, LOWER_UPPER_DIGIT_SPECIAL, FIRST_TO_LOWER_SPECIAL, ALL_TO_LOWER_SPECIAL
    };
```

I coded a unit test, as follows, which pass
```java
@Test
public void testEncoding() {
    String str = "你好，世界";
    MetaString metaString = Encoders.encodePackage(str);
    MetaStringBytes metaStringBytes = new MetaStringBytes(metaString);
    String decode = metaStringBytes.decode(Encoders.PACKAGE_DECODER);
    Assert.assertEquals(str, decode);
}
```

> And the key here is that MetaStringBytes can't distinguish whether the data is package or classname, so the decode need use the whole encoding set

`MetaStringBytes#decode` uses the encoding stored in `MetaStringBytes#hashCode & 0x111`, which means that if the encoding is successful, the decoding will be successful. It should have nothing to do with the encoding set.

Reply:
And `MetaStringEncoder` supports UTF8 by default, even if my `EncodingSet` does not specify UTF8

Reply:
The encoding set in `MetaStringBytes#hashCode & 0x111` use whole `org.apache.fury.meta.MetaString.Encoding.values()`. But the `Encoders.encodePackage` only use `UTF_8, ALL_TO_LOWER_SPECIAL, LOWER_UPPER_DIGIT_SPECIAL`. The inconsistency may introduce some error

Reply:
`MetaStringBytes#hashCode & 0b111` only controls the range of the encoding set, not the encoding set to be used.

> The inconsistency may introduce some error

Could you elaborate on this? I'm not sure what will cause the inconsistency.

Reply:
I misunderstood it. `Encoders.encodePkgName` use `UTF_8, ALL_TO_LOWER_SPECIAL, LOWER_UPPER_DIGIT_SPECIAL` encoding. The `ClassDefEncoder` use index in `UTF_8, ALL_TO_LOWER_SPECIAL, LOWER_UPPER_DIGIT_SPECIAL` to store encoding. But `getOrCreateMetaStringBytes` uses the encoding in returned `MetaString` instead, so it will use `metaString.getEncoding().getValue()` as the index, and there won't be inconsistency when decoding the meta string